### PR TITLE
Provide custom grpc channel

### PR DESCRIPTION
**Pull request by JonatanMartens**
Add ability to provide custom grpc channel to worker/client

## Changes

- grpc channel is now provided by the user instead of being created in pyzeebe.
- Remove all credentials classes

## API Updates

- ZeebeWorker now receives a grpc channel in the constructor
- ZeebeClient now receives a grpc channel in the constructor

### New Features 

- Users can now create their own custom grpc channels instead of having to rely on pyzeebe

### Deprecations 

- Remove credentials classes. These are now unnecessary

### Enhancements

none

## Checklist

- [x] Unit tests
- [x] Documentation

## References

Implements #185

### CODE REVIEW

File: tests/unit/channel/insecure_channel_test.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class TestCreateInsecureChannel:
+    @pytest.fixture(autouse=True)
+    def insecure_channel_mock(self, aio_grpc_channel: grpc.aio.Channel):
+        with patch("grpc.aio.insecure_channel", return_value=aio_grpc_channel) as mock:
+            yield mock
+
+    def test_returns_aio_grpc_channel(self):
+        channel = create_insecure_channel()
```
Comment (kbakk) [inline tests/unit/channel/insecure_channel_test.py:16]: We are missing test for overriding host/port

File: tests/unit/channel/secure_channel_test.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+class TestCreateSecureChannel:
+    @pytest.fixture(autouse=True)
+    def secure_channel_mock(self, aio_grpc_channel: grpc.aio.Channel):
+        with patch("grpc.aio.secure_channel", return_value=aio_grpc_channel) as mock:
+            yield mock
+
+    def test_returns_grpc_channel(self):
+        channel = create_secure_channel()
+
+        assert isinstance(channel, grpc.aio.Channel)
```
Comment (kbakk) [inline tests/unit/channel/secure_channel_test.py:18]: Should test that host/port is being set correctly (can maybe do that in test_calls_using_default_grpc_options)

Should test overriding host/port.

**Review by kbakk — COMMENTED** [review]
Well done! Some comments on testing.

File: examples/worker.py [hunk 3]
```diff
@@ -12,15 +13,22 @@ async def example_logging_task_decorator(job: Job) -> Job:
 
 
 # Will use environment variable ZEEBE_ADDRESS or localhost:26500 and NOT use TLS
-worker = ZeebeWorker()
+# create_insecure_channel returns a grpc.aio.Channel instance. If needed you
+# can build one on your own
+grpc_channel = create_insecure_channel()
+worker = ZeebeWorker(grpc_channel)
```
Comment (kbakk) [inline examples/worker.py:19]: Could be an idea to show usage with port/host specified

File: pyzeebe/channel/camunda_cloud_channel.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    )
+
+    return grpc.aio.secure_channel(
+        f"{cluster_id}.zeebe.camunda.io:443",
+        channel_credentials,
+        options=get_channel_options(channel_options),
+    )
+
+
+def create_camunda_cloud_credentials(
```
Comment (kbakk) [inline pyzeebe/channel/camunda_cloud_channel.py:45]: prefix with _ to indicate private function? (same with get_access_token)?

File: pyzeebe/channel/insecure_channel.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+from pyzeebe.channel.channel_options import get_channel_options
+from pyzeebe.channel.utils import create_address
+
+
+def create_insecure_channel(
+    hostname: Optional[str] = None,
+    port: Optional[int] = None,
+    channel_options: Optional[Dict] = None
+) -> grpc.aio.Channel:
```
Comment (kbakk) [inline pyzeebe/channel/insecure_channel.py:13]: Consider simply `create_channel(secure=False)`? Not sure what the advantage is with separated `create_secure_channel`/`create_insecure_channel`

File: pyzeebe/channel/utils.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+from typing import Optional
+
+
+def create_address(
+    hostname: Optional[str] = None,
+    port: Optional[int] = None,
+) -> str:
+    if hostname or port:
+        return f"{hostname or 'localhost'}:{port or 26500}"
+    return os.getenv("ZEEBE_ADDRESS", "localhost:26500")
```
Comment (kbakk) [inline pyzeebe/channel/utils.py:11]: Can we move `localhost:26500` and `26500` to constants - to make clearer why these are used. 

```py
DEFAULT_HOSTNAME = "localhost"
DEFAULT_PORT = 26500
DEFAULT_ADDRESS = f"{DEFAULT_HOSTNAME}:{DEFAULT_PORT}"
```

If there are other references under `pyzeebe/`, these should maybe be imported from one file (single source of truth).

File: pyzeebe/client/client.py [hunk 7]
```diff
@@ -2,28 +2,26 @@
...
 
 
 class ZeebeClient(object):
     """A zeebe client that can connect to a zeebe instance and perform actions."""
 
-    def __init__(self, hostname: str = None, port: int = None, credentials: BaseCredentials = None,
-                 secure_connection: bool = False, max_connection_retries: int = 10):
+    def __init__(
+        self, grpc_channel: grpc.aio.Channel, max_connection_retries: int = 10
+    ):
```
Comment (kbakk) [inline pyzeebe/client/client.py:13]: Started to use Black?

**Comment by kbakk** [comment]
Thanks for the continued maintenance and improvement of this library! Well done with the efforts here. No major pain points discovered during review. I'm not 100 % onboard with dependency injecting the GRPC client since it feels like exposing a low-level detail at a high level, and it adds some repetition for the users. It's only done once per client/worker, so it's not that much extra work needed.

File: pyzeebe/client/client.py (hunk 7 above)
Comment (JonatanMartens) [reply pyzeebe/client/client.py:13]: Yes, I'm considering using it in pyzeebe :)

File: pyzeebe/channel/insecure_channel.py (hunk 5 above)
Comment (JonatanMartens) [reply pyzeebe/channel/insecure_channel.py:13]: You can provide your own `grpc.ChannelCredentials` when using `create_secure_channel`. This way we avoid confusion where users may try:

```python
create_secure_channel(secure=False, channel_credentials=some_credendtials)
```

**Comment by JonatanMartens** [comment]
> I'm not 100 % onboard with dependency injecting the GRPC client since it feels like exposing a low-level detail at a high level, and it adds some repetition for the users

Definitely agree, but I don't see a way to avoid these disadvantages without taking away the customizability introduced here.

**Pull request merged**